core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -358,58 +361,118 @@ private static BulkItemResultHolder executeUpdateRequest(UpdateRequest updateReq
         return new BulkItemResultHolder(updateResponse, updateOperationResult, replicaRequest);
     }
 
-    static boolean shouldExecuteReplicaItem(final BulkItemRequest request, final int index) {
+    /** Modes for executing item request on replica depending on corresponding primary execution result */
+    public enum ReplicaItemExecutionMode {
+
+        /**
+         * When primary execution succeeded
+         */
+        NORMAL,
+
+        /**
+         * When primary execution failed before sequence no was generated
+         * or primary execution was a noop (only possible when request is originating from pre-6.0 nodes)
+         */
+        NOOP,
```

Comment:
Can you add an assert that will fail on 7.0.0 that we should remove the NOOP and all handling code for it?